common/src/main/java/org/conscrypt/NativeSsl.java

```diff
@@ -611,14 +611,18 @@ protected final void finalize() throws Throwable {
      * A utility wrapper that abstracts operations on the underlying native BIO instance.
      */
     final class BioWrapper {
-        private long bio;
+        private volatile long bio;
 
         private BioWrapper() throws SSLException {
             this.bio = NativeCrypto.SSL_BIO_new(ssl, NativeSsl.this);
         }
 
         int getPendingWrittenBytes() {
-            return NativeCrypto.SSL_pending_written_bytes_in_BIO(bio);
+            if (bio != 0) {
+                return NativeCrypto.SSL_pending_written_bytes_in_BIO(bio);
```

Comment:
I think you should store the bio in a local variable  first for two reasons:

- performance
- it may be possible otherwise it change in between multiple volatile loads of the bio

Reply:
Got it...